batch/batch/cloud/gcp/worker/disk.py

```diff
@@ -94,19 +76,41 @@ async def _attach(self):
             await self.compute_client.attach_disk(
                 f'/zones/{self.zone}/instances/{self.instance_name}/attachDisk', json=config
             )
-            self._attached = True
+            self.attached = True
 
     async def _detach(self):
         async with LoggingTimer(f'detaching disk {self.name} from {self.instance_name}'):
-            if self._attached:
+            if self.attached:
                 await self.compute_client.detach_disk(
                     f'/zones/{self.zone}/instances/{self.instance_name}/detachDisk', params={'deviceName': self.name}
                 )
+                self.attached = False
 
     async def _delete(self):
         async with LoggingTimer(f'deleting disk {self.name}'):
-            if self._created:
+            if self.created:
                 await self.compute_client.delete_disk(f'/zones/{self.zone}/disks/{self.name}')
+                self.created = False
 
     def __str__(self):
         return self.name
+
+
+class GCPDiskManager(CloudDiskManager):
+    def __init__(self, project: str, zone: str, max_disks: int = 128):
+        self.project = project
+        self.zone = zone
+        self.disk_slots = asyncio.Semaphore(max_disks)
+
+    async def new_disk(self, instance_name: str, disk_name: str, size_in_gb: int, mount_path: str) -> GCPDisk:
+        await self.disk_slots.acquire()
+        return GCPDisk(disk_name, self.zone, self.project, instance_name, size_in_gb, mount_path)
+
+    async def delete_disk(self, disk: GCPDisk):  # type: ignore[override]
+        try:
+            await disk.delete()
+        finally:
+            try:
+                self.disk_slots.release()
+            finally:
+                await disk.close()
```

Comment:
I don't see any reason to delete the disk before release but not close it. I think we should do both before releasing the semaphore.